src/librustc_incremental/persist/save.rs

```diff
@@ -98,38 +101,37 @@ fn save_in<F>(sess: &Session, opt_path_buf: Option<PathBuf>, encode: F)
     }
 }
 
-pub fn encode_dep_graph<'a, 'tcx>(hcx: &mut HashContext<'a, 'tcx>,
-                                  builder: &mut DefIdDirectoryBuilder,
-                                  query: &DepGraphQuery<DefId>,
-                                  encoder: &mut Encoder)
-                                  -> io::Result<()> {
-    let (nodes, edges) = (query.nodes(), query.edges());
-
-    // Create hashes for inputs.
-    let hashes = nodes.iter()
-        .filter_map(|dep_node| {
-            hcx.hash(dep_node)
-                .map(|(_, hash)| {
-                    let node = builder.map(dep_node);
-                    SerializedHash {
-                        node: node,
-                        hash: hash,
-                    }
-                })
-        })
-        .collect();
+pub fn encode_dep_graph(preds: &Predecessors,
+                        builder: &mut DefIdDirectoryBuilder,
+                        encoder: &mut Encoder)
+                        -> io::Result<()> {
+    // Create a flat list of (Input, WorkProduct) edges for
+    // serialization.
+    let mut edges = vec![];
+    for (&target, sources) in &preds.inputs {
+        match *target {
+            DepNode::MetaData(_) => continue, // see encode_metadata_hashes instead
```

Comment:
> Also, with some additional machinery, similar to what we do for object files, we could integrate the metadata-object into the dependency graph too 

Yes, I've thought about this. We have basically all the data we need, I think, but we have to be able to save/reload the meta-data. One problem with this is that it includes def-ids which may have changed and would have to be retraced. (This is why we don't just hash the raw bytes of the metadata.) I guess we could encode meta-data using the same "def-path-index" system, so that we do the work of retracing when we load the meta-data, instead of when we write it. In short, we have to think about this a bit, but it's all definitely doable.